cuda_core/examples/memory_ops.py

```diff
@@ -0,0 +1,138 @@
+# Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES. ALL RIGHTS RESERVED.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+# ################################################################################
+#
+# This demo illustrates:
+#
+#   1. How to use different memory resources to allocate and manage memory
+#   2. How to copy data between different memory types
+#   3. How to use DLPack to interoperate with other libraries
+#
+# ################################################################################
+
+import sys
+
+import cupy as cp
+import numpy as np
+
+from cuda.core.experimental import (
+    Device,
+    DeviceMemoryResource,
+    LaunchConfig,
+    LegacyPinnedMemoryResource,
+    Program,
+    ProgramOptions,
+    launch,
+)
+
+if np.__version__ < "2.1.0":
+    print("This example requires NumPy 2.1.0 or later", file=sys.stderr)
+    sys.exit(0)
+
+# Kernel for memory operations
+code = """
+extern "C"
+__global__ void memory_ops(float* device_data,
+                          float* pinned_data,
+                          size_t N) {
+    const unsigned int tid = threadIdx.x + blockIdx.x * blockDim.x;
+    if (tid < N) {
+        // Access device memory
+        device_data[tid] = device_data[tid] + 1.0f;
+
+        // Access pinned memory (zero-copy from GPU)
+        pinned_data[tid] = pinned_data[tid] * 3.0f;
+    }
+}
+"""
+
+dev = Device()
+dev.set_current()
+stream = dev.create_stream()
+# tell CuPy to use our stream as the current stream:
+cp.cuda.ExternalStream(int(stream.handle)).use()
+
+# Compile kernel
+arch = "".join(f"{i}" for i in dev.compute_capability)
+program_options = ProgramOptions(std="c++17", arch=f"sm_{arch}")
+prog = Program(code, code_type="c++", options=program_options)
+mod = prog.compile("cubin")
+kernel = mod.get_kernel("memory_ops")
+
+# Create different memory resources
+device_mr = DeviceMemoryResource(dev.device_id)
```

Comment:
I will test this tomorrow but I suspect this is what happened on Windows:
```suggestion
try:
    device_mr = DeviceMemoryResource(dev.device_id)
except Exception as e:
    print("This example requires a driver supporting stream-ordered memory allocator", file=sys.stderr)
    cp.cuda.Stream.null.use()  # reset
    sys.exit(0)
```

Reply:
One alternative to run the same code on Windows is to switch to the per-device MR from using an independent MR, with which we hide the platform difference:
```suggestion
device_mr = dev.memory_resource
```

Reply:
It should not be identical because it's neither how the driver works nor in users' best interest to unify.

At the driver level, users are free to create as many `cudaMempool_t` as they like, because each pool could be configured differently. So the cuda.core semantics of `dev.memory_resource` reads: "Give me the memory resource that wraps the device's current (or default, if no one touches it) mempool." However, users can create a side `DeviceMemoryResource` instance wrapping a new instance of `cudaMempool_t` (say by passing a pool config or by wrapping a foreign `cudaMempool_t` pointer) and allocate from there, instead of from `dev.memory_resource`, just like how they could do the same at the C level with raw driver APIs. In fact, #446 demonstrates such a use case: Allocating a new mempool for IPC purposes (be careful when reading the diff there, the branch has staled and deviated from the main branch for a while 😅). So if we cache and make them the same, we lose this capability entirely, and misaligns with C (raw driver) and C++ (cudax) designs.

There is one thing we do need to solve, which is `Device(0).memory_resource` from thread 0 != `Device(0).memory_resource` from thread 1 when comparing using the Python `__eq__`, despite under the hood they might wrap the same `cudaMempool_t`, due to `Device` being per-thread instances. But we can fix it in another PR.

Reply:
Okay, that makes sense - thanks for clarifying!

Reply:
I replied below